solvers/wscd.py

```diff
@@ -0,0 +1,197 @@
+
+
+@njit
+def sparse_cd(
+        X_data, X_indices, X_indptr, y, lmbd, gamma, lipschitz, n_iter, winit,
+        tol=1e-5):
+    n_features = len(X_indptr) - 1
+    n_samples = len(y)
+    all_feats = np.arange(n_features)
+    w = np.zeros(n_features)
+    R = np.copy(y)
+    for _ in range(n_iter):
+        for j in range(n_features):
+            if lipschitz[j]:
+                old = w[j]
+                XjtR = 0.0
+                for ind in range(X_indptr[j], X_indptr[j + 1]):
+                    XjtR += X_data[ind] * R[X_indices[ind]]
+
+                w[j] = prox_mcp(w[j] + XjtR / (lipschitz[j] * n_samples),
+                                lmbd / lipschitz[j], gamma * lipschitz[j])
+                diff = old - w[j]
+                if diff != 0:
+                    for ind in range(X_indptr[j], X_indptr[j + 1]):
+                        R[X_indices[ind]] += diff * X_data[ind]
+        subdiff_dist = subdiff_distance(w, R, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            return w
+    return w
+
+
+@njit
+def wscd(X, y, lmbd, gamma, lipschitz, n_iter, n_iter_outer, pruning=True,
+         tol=1e-8, sparsity=False):
+
+    n_samples, n_features = X.shape
+    nb_feat_init = 10
+    nb_feat_2_add = 30
+    ind = np.argsort(-np.abs(X.T.dot(y)))[:nb_feat_init]
+
+    all_feats = np.arange(n_features)
+    w_init = np.zeros((nb_feat_init))
+    w = np.zeros((n_features))
+
+    for i in range(n_iter_outer):
+        Xaux = X[:, ind]
+
+        if sparsity:
+            pass
+            # TODO a proper call to the appropriate sparse function
+        else:
+            w_inter = cd(Xaux, y, lmbd, gamma, lipschitz[ind], n_iter, w_init)
+        # pruning
+        if pruning:
+            nnz = (w_inter != 0)
+            w_inter = w_inter[nnz]
+            ind = ind[nnz]
+            Xaux = Xaux[:, nnz]
+
+        res = y - Xaux @ w_inter
+        grad = - X.T @ res / n_samples
+
+        subdiff_dist = subdiff_distance(w, grad, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            w[ind] = w_inter
+            return w
+        grad[ind] = 0
+        candidate = np.argsort(-np.abs(grad))
+        # TODO use argpartition when numba implems is available
+
+        ind = np.hstack((ind, candidate[:nb_feat_2_add]))
+        w_init = np.hstack((w_inter, np.zeros(nb_feat_2_add)))
+    w = np.zeros(n_features)
+    w[ind] = w_init
+
+    return w
+
+
+class Solver(BaseSolver):
+    name = "working_set_cd"
+    install_cmd = "conda"
+    requirements = ["numba"]
+
+    def set_objective(self, X, y, lmbd, gamma):
+        self.X, self.y = X, y
+        self.lmbd, self.gamma = lmbd, gamma
+        self.n_iter_outer = 10
+        # Make sure we cache the numba compilation.
+        self.run(1)
+
+    def run(self, n_iter):
+        X = self.X
+        if sparse.issparse(X):
+            lipschitz = norm_col_sparse(X.data, X.indptr) ** 2 / len(self.y)
+            sparsity = True
```

Comment:
sparsity can be computed inside wscd, in order to factorize the code and remove one variable from the function signature